### PR TITLE
add NetworkManager remote control via agent

### PR DESCRIPTION
**Pull request by jluebbe**
**Description**
This adds support for controlling an exported NetworkInterface remotely using NetworkManager over an agent connection.

Via NetworkManager, we can configure and inspect Ethernet and WiFi (client&AP) interfaces. This is especially useful for testing WiFi interfaces on the DUT.

**Checklist**
- [x] Documentation for the feature
- [ ] Tests for the feature 
- [x] The arguments and description in doc/configuration.rst have been updated
- [ ] CHANGES.rst has been updated
- [x] PR has been tested

### CODE REVIEW

File: labgrid/driver/networkinterfacedriver.py [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+        super().__attrs_post_init__()
+        self.wrapper = None
+
+    def on_activate(self):
+        if isinstance(self.iface, RemoteNetworkInterface):
+            host = self.iface.host
+        else:
+            host = None
+        self.wrapper = AgentWrapper(host)
+        self.proxy = self.wrapper.load('network_interface')
```
Comment (Emantor) [inline labgrid/driver/networkinterfacedriver.py:28]: Maybe `agent` instead of `proxy`?

Comment (jluebbe) [reply labgrid/driver/networkinterfacedriver.py:28]: This is a proxy object for the agent and we already use 'proxy' in the other drivers. So I'd tend to keep it for consistency.

**Comment by jluebbe** [comment]
I've added some documentation.

**Comment by jluebbe** [comment]
I've rebased it on master.

File: labgrid/util/agents/network_interface.py [hunk 2]
```diff
@@ -0,0 +1,359 @@
...
+import gi
+gi.require_version('NM', '1.0')
+from gi.repository import GLib, Gio, NM
+
+# ensure all wrapper objects for Settings types are created
+for name in dir(NM):
+    if name.startswith('Setting'):
+        getattr(NM, name)
+
+import threading
```
Comment (Bastian-Krause) [inline labgrid/util/agents/network_interface.py:15]: This is an import from the stdlib, so move it up?

File: labgrid/util/agents/network_interface.py [hunk 3]
```diff
@@ -0,0 +1,359 @@
...
+        nm.add_and_activate_connection_async(
+            con,
+            self._nm_dev,
+            None,  # specific_object
+            None,  # cancellable
+            cb,
+            None,
+        )
+
+        active_con = future.wait()
```
Comment (Bastian-Krause) [inline labgrid/util/agents/network_interface.py:159]: Will `active_con` be used in the future?

File: labgrid/util/agents/network_interface.py [hunk 4]
```diff
@@ -0,0 +1,359 @@
...
+
+    def wait_state(self, expected, timeout):
+        res, out_value, err_token = NM.utils_enum_from_str(NM.DeviceState, expected)
```
Comment (Bastian-Krause) [inline labgrid/util/agents/network_interface.py:162]: `err_token` seems to be unused.

File: labgrid/util/agents/network_interface.py [hunk 5]
```diff
@@ -0,0 +1,359 @@
+import sys
+import uuid
```
Comment (Bastian-Krause) [inline labgrid/util/agents/network_interface.py:2]: Unused import.

File: labgrid/util/agents/network_interface.py [hunk 6]
```diff
@@ -0,0 +1,359 @@
...
+            except Exception as e:
+                future.set(e)
+
+        self._nm_dev.request_scan_async(
+            None,  # cancellable
+            cb,
+            None,
+        )
+
+        res = future.wait()
```
Comment (Bastian-Krause) [inline labgrid/util/agents/network_interface.py:257]: Will `res` be used in the future?

File: labgrid/util/agents/network_interface.py [hunk 7]
```diff
@@ -0,0 +1,359 @@
...
+    def set(self, result):
+        assert self._result is None
+        self._result = result
+        self._event.set()
+
+    def wait(self):
+        self._event.wait()
+        if isinstance(self._result, Exception):
+            raise self._result
+        else:
```
Comment (Bastian-Krause) [inline labgrid/util/agents/network_interface.py:31]: We can drop the else (after raise).

File: labgrid/util/agents/network_interface.py [hunk 8]
```diff
@@ -0,0 +1,359 @@
+import sys
+import uuid
+import socket
+from time import monotonic, sleep
+
+import gi
+gi.require_version('NM', '1.0')
+from gi.repository import GLib, Gio, NM
```
Comment (Bastian-Krause) [inline labgrid/util/agents/network_interface.py:8]: `Gio` is unused.

Comment (Bastian-Krause) [inline labgrid/util/agents/network_interface.py:8]: Does it make sense to add `pygobject` to one of the requirements.txt files?

**Review by Bastian-Krause — CHANGES_REQUESTED** [review]

File: labgrid/util/agents/network_interface.py [hunk 9]
```diff
@@ -0,0 +1,359 @@
...
+
+        con.delete_async(
+            None,  # cancellable
+            cb,
+            None,
+        )
+
+    def get_settings(self):
+        lg_con = nm.get_connection_by_id("labgrid-{}".format(self._interface))
+        if lg_con:
```
[the remark this answers is not in the document]
Comment (jluebbe) [reply labgrid/util/agents/network_interface.py:121]: Returning None if there is no configured connection is fine.

File: labgrid/util/agents/network_interface.py (hunk 3 above)
Comment (jluebbe) [reply labgrid/util/agents/network_interface.py:159]: No, I'll add a comment.

File: labgrid/util/agents/network_interface.py (hunk 5 above)
Comment (jluebbe) [reply labgrid/util/agents/network_interface.py:2]: Removed.

File: labgrid/util/agents/network_interface.py (hunk 6 above)
Comment (jluebbe) [reply labgrid/util/agents/network_interface.py:257]: I've added a comment as above.

File: labgrid/util/agents/network_interface.py (hunk 8 above)
Comment (jluebbe) [reply labgrid/util/agents/network_interface.py:8]: No, as it cannot be installed in the virtualenv and must come from the host.

File: labgrid/util/agents/network_interface.py [hunk 10]
```diff
@@ -0,0 +1,359 @@
...
+            if self._nm_dev.get_state() == expected:
+                break
+        else:
+            raise TimeoutError("state is '{}' instead of '{}'".format(
+                self._nm_dev.get_state().value_nick,
+                expected.value_nick))
+
+    def disable(self):
+        lg_con = nm.get_connection_by_id("labgrid-{}".format(self._interface))
+        if lg_con:
```
[the remark this answers is not in the document]
Comment (jluebbe) [reply labgrid/util/agents/network_interface.py:179]: If there is no connection, this should become a no-op.

File: labgrid/util/agents/network_interface.py [hunk 11]
```diff
@@ -0,0 +1,359 @@
...
+        )
+
+    def get_settings(self):
+        lg_con = nm.get_connection_by_id("labgrid-{}".format(self._interface))
+        if lg_con:
+            return dict(lg_con.to_dbus(NM.ConnectionSerializationFlags.ALL))
+
+    def get_active_settings(self):
+        active_con = self._nm_dev.get_active_connection()
+        if active_con:
```
[the remark this answers is not in the document]
Comment (jluebbe) [reply labgrid/util/agents/network_interface.py:126]: Returning None if there is no active connection is intended.

**Review by Emantor — APPROVED** [review]

**Pull request merged**